### Partie 1 – Explorer les données 

##### 1) Charger les données CSV 

In [2]:
import numpy as np
import pandas as pd 

In [3]:
df = pd.read_csv("../data/smart_building_raw.csv")
print(df)

     id_mesure                 date batiment      type_batiment zone  \
0         1174  2025-02-13 06:00:00       B8           Entrepôt    A   
1         1275  2025-03-10 12:00:00       B7             Bureau    D   
2         1493  2025-05-04 00:00:00       B5  Centre commercial    B   
3         1073  2025-01-19 00:00:00       B6         Université    D   
4         1454  2025-04-24 06:00:00       B7             Bureau    D   
..         ...                  ...      ...                ...  ...   
502       1107  2025-01-27 12:00:00       B2              École    C   
503       1271  2025-03-09 12:00:00       B3            Hôpital    B   
504       1349  2025-03-29 00:00:00       B5  Centre commercial    B   
505       1436  2025-04-19 18:00:00       B7             Bureau    C   
506       1103  2025-01-26 12:00:00       B2              École    D   

     temperature  humidite     co2  occupation  consommation_kwh  \
0           23.0      40.0   851.0        26.0             138.4   

##### 2) Afficher les premières lignes du dataset

In [4]:
print(df.head())

   id_mesure                 date batiment      type_batiment zone  \
0       1174  2025-02-13 06:00:00       B8           Entrepôt    A   
1       1275  2025-03-10 12:00:00       B7             Bureau    D   
2       1493  2025-05-04 00:00:00       B5  Centre commercial    B   
3       1073  2025-01-19 00:00:00       B6         Université    D   
4       1454  2025-04-24 06:00:00       B7             Bureau    D   

   temperature  humidite     co2  occupation  consommation_kwh  \
0         23.0      40.0   851.0        26.0             138.4   
1         28.0      69.0   538.0         0.0             170.2   
2         19.5      52.3  1198.0        64.0             149.8   
3         20.0      58.4  1014.0        30.0             141.6   
4         26.8      70.5   628.0        42.0             228.3   

  mode_climatisation etat_systeme jour_semaine alerte  
0                Eco       Normal        Jeudi    Non  
1                Eco       Normal        Lundi    Non  
2             

##### 3) Afficher les dernières lignes du dataset 

In [5]:
print(df.tail())

     id_mesure                 date batiment      type_batiment zone  \
502       1107  2025-01-27 12:00:00       B2              École    C   
503       1271  2025-03-09 12:00:00       B3            Hôpital    B   
504       1349  2025-03-29 00:00:00       B5  Centre commercial    B   
505       1436  2025-04-19 18:00:00       B7             Bureau    C   
506       1103  2025-01-26 12:00:00       B2              École    D   

     temperature  humidite     co2  occupation  consommation_kwh  \
502         29.8      66.5   778.0        48.0             243.5   
503         21.1      66.3  3900.0        53.0             145.8   
504         21.6      61.9   959.0        33.0             100.2   
505         21.2      54.5  1066.0        38.0             150.9   
506          NaN      64.8   584.0        29.0             163.4   

    mode_climatisation etat_systeme jour_semaine alerte  
502             Normal       Normal        Lundi    Non  
503                Eco       Alerte     Di

##### 4) Combien d'observations contient le dataset 

In [6]:
df.shape[0]
print ("Le nombre d'observation que contient le dataset est : 507")

Le nombre d'observation que contient le dataset est : 507


##### 5) Combien de variables possède le dataset 

In [7]:
df.columns.tolist()

['id_mesure',
 'date',
 'batiment',
 'type_batiment',
 'zone',
 'temperature',
 'humidite',
 'co2',
 'occupation',
 'consommation_kwh',
 'mode_climatisation',
 'etat_systeme',
 'jour_semaine',
 'alerte']

##### 6) Identifier les variables numériques 

In [8]:
var_num = df.select_dtypes(include=np.number).columns.tolist()
print(var_num)

['id_mesure', 'temperature', 'humidite', 'co2', 'occupation', 'consommation_kwh']


##### 7) Identifier les variables catégorielles 

In [9]:
print("les variables categorielles sont :", df.select_dtypes(include="object").columns.tolist())

les variables categorielles sont : ['date', 'batiment', 'type_batiment', 'zone', 'mode_climatisation', 'etat_systeme', 'jour_semaine', 'alerte']


C:\Users\dell\AppData\Local\Temp\ipykernel_5920\3312097636.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  print("les variables categorielles sont :", df.select_dtypes(include="object").columns.tolist())


##### 8) Identifier les dates 

In [10]:
print("les variables de type date:\n", df['date'])

les variables de type date:
 0      2025-02-13 06:00:00
1      2025-03-10 12:00:00
2      2025-05-04 00:00:00
3      2025-01-19 00:00:00
4      2025-04-24 06:00:00
              ...         
502    2025-01-27 12:00:00
503    2025-03-09 12:00:00
504    2025-03-29 00:00:00
505    2025-04-19 18:00:00
506    2025-01-26 12:00:00
Name: date, Length: 507, dtype: str


##### 9) Identifier les identifiants

In [11]:
print("les identifiants sont:\n", df['id_mesure'])

les identifiants sont:
 0      1174
1      1275
2      1493
3      1073
4      1454
       ... 
502    1107
503    1271
504    1349
505    1436
506    1103
Name: id_mesure, Length: 507, dtype: int64


##### 10) Déterminer les statistiques : moyenne, médiane, minimum, maximum, écart-type et quartiles. 

In [15]:
print("les statistiques pour les variables numeriques sont : \n", df.describe())

les statistiques pour les variables numeriques sont : 
          id_mesure  temperature    humidite          co2  occupation  \
count   507.000000   495.000000  496.000000   500.000000  501.000000   
mean   1251.114398    24.154141   57.864113   844.150000   44.850299   
std     144.782769     7.418465   16.026336   582.181386   24.949139   
min    1001.000000   -30.000000  -12.000000    89.000000  -20.000000   
25%    1125.500000    21.600000   49.275000   623.750000   27.000000   
50%    1252.000000    24.000000   57.550000   787.500000   46.000000   
75%    1376.500000    26.000000   65.750000   952.000000   61.000000   
max    1500.000000    96.000000  160.000000  6000.000000  116.000000   

       consommation_kwh  
count        502.000000  
mean         169.069323  
std           53.164294  
min         -100.000000  
25%          136.875000  
50%          169.800000  
75%          202.975000  
max          336.200000  


In [14]:
print("les statistiques pour les variables categorielles sont :\n ", df.describe(include="object"))

les statistiques pour les variables categorielles sont :
                         date batiment type_batiment zone mode_climatisation  \
count                   507      507           503  507                502   
unique                  500        8            16    4                  7   
top     2025-01-30 06:00:00       B1        Bureau    B             Normal   
freq                      2       94           210  143                271   

       etat_systeme jour_semaine alerte  
count           507          502    507  
unique            3            7      2  
top          Normal     Vendredi    Non  
freq            442           75    361  


C:\Users\dell\AppData\Local\Temp\ipykernel_5920\3490907401.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  print("les statistiques pour les variables categorielles sont :\n ", df.describe(include="object"))


##### 11) Y a-t-il des variables potentiellement problématiques 

Oui il y'a plusieurs variables potentiellement problematiques .
--on observe qu'il y'a des valeurs manquantes dans toutes les variables.

id_mesure a un count de 507, alors que les autres colonnes ont entre 495 et 502 valeurs non nulles. Donc il manque environ 12 valeurs en température, 11 en humidité, 7 en CO2, 6 en occupation et 5 en consommation.

--On a probablement la presences de doublons sur id_mesure.Les identifiants vont de 1001 à 1500, soit 500 valeurs possibles, mais il y a 507 lignes. Il y a donc probablement des lignes en double.

--On constate egalement des valeurs aberrantes:

temperature : de -30 à 96 °C, alors que 50 % des valeurs sont entre 21,6 et 26. Ces extrêmes ressemblent à des erreurs de capteur.

humidite : de -12 à 160, alors qu'une humidité en % doit rester entre 0 et 100.

co2 : le max de 6000 est très loin du 75e percentile (952), et l'écart-type (582) est énorme par rapport à la médiane. Le min de 89 est aussi suspect, car l'air extérieur est déjà autour de 400 ppm.

occupation : le min de -20 est impossible. car on ne peut pas avoir un nombre negatif de personnes.

consommation_kwh : le min de -100 est impossible (une consommation ne peut pas être négative).

##### 12) Pour les données incohérentes a) rechercher des valeurs telles que humidité < 0 

In [21]:
print("les valeurs telles que l'humidite est inferieur a 0 sont:\n", df[df["humidite"] < 0])

les valeurs telles que l'humidite est inferieur a 0 sont:
      id_mesure                 date batiment type_batiment zone  temperature  \
281       1126  2025-02-01 06:00:00       B2         École    D         24.8   
335       1036  2025-01-09 18:00:00       B4        Bureau    D         21.2   
366       1216  2025-02-23 18:00:00       B3       Hôpital    B         23.2   

     humidite    co2  occupation  consommation_kwh mode_climatisation  \
281      -5.0  759.0        57.0             225.2              Boost   
335      -8.0  160.0        25.0             120.2             Normal   
366     -12.0  702.0        41.0             119.8             Normal   

    etat_systeme jour_semaine alerte  
281       Normal       Samedi    Non  
335       Alerte        Jeudi    Non  
366       Normal     Dimanche    Non  


#####  12) Pour les données incohérentes b) rechercher des valeurs telles que humidité > 100

In [22]:
print("les valeurs telles que l'humidite est superieur a 100 sont:\n", df[df["humidite"] > 100])

les valeurs telles que l'humidite est superieur a 100 sont:
      id_mesure                 date batiment      type_batiment zone  \
59        1246  2025-03-03 06:00:00       B5  Centre commercial    A   
103       1016  2025-01-04 18:00:00       B6         Université    B   
128       1156  2025-02-08 18:00:00       B1             Bureau    A   
160       1186  2025-02-16 06:00:00       B8           Entrepôt    B   
268       1276  2025-03-10 18:00:00       B1             Bureau    D   
327       1066  2025-01-17 06:00:00       B2              École    C   
342       1096  2025-01-24 18:00:00       B8           Entrepôt    A   

     temperature  humidite    co2  occupation  consommation_kwh  \
59          24.2     108.0  813.0        18.0             -15.0   
103         26.9     145.0  670.0        28.0             175.9   
128          NaN     160.0  941.0        37.0             149.8   
160         24.9     125.0    NaN        18.0             123.5   
268         26.0     140.0 